contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,216 @@
+    // Version returns the runtime name, runtime version, and runtime API version.
+    rpc Version(VersionRequest) returns (VersionResponse) {}
+    …….
+    …….
+    //  RuntimeConfigInfo returns the configuration details of the runtime.
+    rpc RuntimeConfigInfo(RuntimeConfigInfoRequest) returns (RuntimeConfigInfoResponse) {}
+}
+// LinuxIDMapping represents a single user namespace mapping in Linux.
+message LinuxIDMapping {
+   // container_id is the starting id for the mapping inside the container.
+   uint32 container_id = 1;
+   // host_id is the starting id for the mapping on the host.
+   uint32 host_id = 2;
+   // size is the length of the mapping.
+   uint32 size = 3;
+}
+
+// UserNamespaceMapping is a user namespace mapping for uids and gids.
+message UserNamespaceMapping {
+   // uid_mappings is an array of user id mappings.
+   repeated LinuxIDMapping uid_mappings = 1;
+   // gid_mappings is an array of group id mappings.
+   repeated LinuxIDMapping gid_mappings = 2;
+}
+
+message UserNamespaceConfig {
+   // is_enabled, if true indicates that user-namespaces are supported and enabled in the container runtime
+   string is_enabled = 1;
+   // node_usernamespace_mapping is the user namespace mapping set for the container runtime
+   UserNSMapping node_usernamespace_mapping = 2;
+}
+message RuntimeConfig {
+    NetworkConfig network_config = 1;
+    UserNamespaceConfig user_namespace_config = 2;
+}
+
+message UpdateRuntimeConfigRequest {
+    RuntimeConfig runtime_config = 1;
+}
+
+message UpdateRuntimeConfigResponse {}
+
+message RuntimeConfigInfoRequest {}
+
+message RuntimeConfigInfoResponse {
+    RuntimeConfig runtime_config = 1
+}
+
+...
+
+// NamespaceOption provides options for Linux namespaces.
+message NamespaceOption {
+	// Network namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped network in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode network = 1;
+	// PID namespace for this container/sandbox.
+	// Note: The CRI default is POD, but the v1.PodSpec default is CONTAINER.
+	// The kubelet's runtime manager will set this to CONTAINER explicitly for v1 pods.
+	// Namespaces currently set by the kubelet: POD, CONTAINER, NODE
+	NamespaceMode pid = 2;
+	// IPC namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped IPC in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode ipc = 3;
+	// User namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped user namespace in the
+	// Kubernetes API. The container runtime has to run with user namespace enabled,
+	// and NODE then forces the use of host user namespace.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceModeValue user = 4;
+}
+
+// NamespaceModeValue is the wrapper of NamespaceMode, to allow non-POD default.
+message NamespaceModeValue {
```

Comment:
The use of `NamespaceModeValue` wrapper for `user` seems strange. Ideally we might want to go with just `NamespaceMode user = 4;` like we do for `network`, `pid`, and `ipc`. The problem is in the default. Ideally we'd like to go with `POD` ... but with existing docker Container Runtimes without `--userns-remap` specified, we always get no remapping (== effectively `NODE`) anyway.

Do we have some options to resolve this issue?

Reply:
Thats what we are trying to solve here by using this wrapper `NamespaceModeValue`. in proto3, sub-messages can be set to `nil`.

Reply:
Shouldn't the default be set by the kubelet?

Reply:
It's be nice to be able to start setting the default from now on. But what should the default be? `NODE`, and `POD` only when we detect that the Container Runtime can support it and will do it for us (like docker with `--userns-remap` enabled), but not when `NODE` is requested by `ExperimentalHostUserNamespaceDefaulting` gate or `hostUserNamespace`?

Reply:
@adelton @tallclair By default kubelet will not set this variable and thus keeping behavior same as it is today when this member does not exist. Runtime will see a `nil` value. `POD` will be set by kubelet if `NodeUserNamespace` is enabled and `NODE` will be set if `NodeUserNamespace` is enabled but a pod wants to escape user-namespace.
Runtime will have clear indication about each condition without ambiguity.

Reply:
Well, and the question is, could we say that from now one, not having the member in the message will be considered the default for `NamespaceMode` which is `POD`, and we immediately change the code to always set it to `NODE` because after all, that's what it semantically means today.

And the for alpha / 1.11, we will add the code to set the `POD` when we detect `NodeUserNamespace` and escape to `NODE` is not requested nor allowed.

What I'm aiming for with these thoughts is to make the `user` consistent with the existing `network`, `pid`, and `ipc`, actually defining the behaviour, without that `nil` state when we say "it behaves like it always did, except we actually don't know how the Pod will be run".

Reply:
Can't the kubelet just default to POD and set to NODE when trying to use host user namespace. The runtime can ignore the value when it has not been configured with user namespaces. We don't need a wrapper if we are okay with that.

Reply:
This too is fine with me. @adelton @tallclair thoughts?

Reply:
Right, that will work fine as well.

I'm not sure we want to declare that the Container Runtime is free to ignore the value (that `POD` default) though. After all, running with `--userns=host` when the kubelet thinks user namespaces are applied means that the payload runs with different security attributes and intended.